Zend/zend_call_stack.c

```diff
@@ -275,6 +275,14 @@ static bool zend_call_stack_get_freebsd_pthread(zend_call_stack *stack)
 		goto fail;
 	}
 
+    error = pthread_attr_getguardsize(&attr, &guard_size);
+    if (error) {
+        return false;
+    }
+
+    addr = (char *)addr + guard_size;
+    max_size -= guard_size;
```

Comment:
From reading the code and testing, addr and size reported by `pthread_attr_getstack()` appear to reflect the usable stack, and they are not affected by the guard size. Based on that, we should not adjust addr and max_size here. But you may have a different experience. Is this required on some FreeBSD version? On Linux this is required in Glibc < 2.8.

In 13.1, the stack is allocated here: https://github.com/freebsd/freebsd-src/blob/releng/13.1/lib/libthr/thread/thr_stack.c#L276. As we can see, the map size is computed from `stacksize + guardsize`. `stacksize` and `guardsize` come directly from the `pthread_attr` (they are just rounded to the nearest page multiple). [`pthread_attr_setstack()`](https://github.com/freebsd/freebsd-src/blob/releng/13.1/lib/libthr/thread/thr_attr.c#L537) does not adjust addr and size with the guardsize, and [`pthread_attr_setguardsize()`](https://github.com/freebsd/freebsd-src/blob/releng/13.1/lib/libthr/thread/thr_attr.c#L425) does not adjust addr and size either. The `get` versions of these functions also report these values directly.

I verified with this program: https://gist.github.com/arnaud-lb/a53eb410b7c26b38fd01eac2cee0467e. The program starts 3 threads: Thread 1 with the default attrs, thread 2 with the guard size set to 1MiB, and thread 3 with the guard size set to 16MiB (larger than the stack size). Each thread then recurses until it reaches the address reported by `pthread_attr_getstack()`. If this address is the actual base of the stack, this should not crash. If this address is in or bellow the guard pages, this should crash. During my tests, this did not crash.

Reply:
The FreeBSD manpage does not help a lot, but it mentions that these functions conform to POSIX.1. The posix spec about pthread_attr_getstack() [says this](https://pubs.opengroup.org/onlinepubs/9699919799/functions/pthread_attr_getstack.html):
> All pages within the stack described by stackaddr and stacksize shall be both readable and writable by the thread.

There could still be a bug or a discrepancy in FreeBSD's implementation, so I wouldn't trust the spec alone, but this matches what I was seeing above.

Reply:
No my mistake even in old releases the guard size is already.

Reply:
No my mistake even in old releases the guard size is already.